src/ssl/context.cpp

```diff
@@ -0,0 +1,287 @@
+/*
+
+Copyright (c) 2020, Paul-Louis Ageneau
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+  * Redistributions of source code must retain the above copyright
+    notice, this list of conditions and the following disclaimer.
+  * Redistributions in binary form must reproduce the above copyright
+    notice, this list of conditions and the following disclaimer in
+    the documentation and/or other materials provided with the distribution.
+  * Neither the name of the author nor the names of its
+    contributors may be used to endorse or promote products derived
+    from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifdef TORRENT_USE_SSL
+
+#include "libtorrent/ssl/context.hpp"
+#include "libtorrent/ssl/stream.hpp"
+#include "libtorrent/settings_pack.hpp"
+
+#define DEBUG_SSL_VERIFY 0
+
+#if DEBUG_SSL_VERIFY
+#include <cstdarg> // for va_list
+#include <cstdio> // for vsnprintf
+#endif
+
+namespace {
+
+#if DEBUG_SSL_VERIFY
+void debug_log(char const* fmt, ...)
+{
+	using namespace std;
+
+	va_list v;
+	va_start(v, fmt);
+
+	char usr[2048];
+	vsnprintf(usr, sizeof(usr), fmt, v);
+	fputs(usr, stderr);
+	fputs("\n", stderr);
+
+	va_end(v);
+}
+#endif // DEBUG_SSL_VERIFY
+
+}
+
+namespace libtorrent {
+namespace ssl {
+
+#ifdef TORRENT_USE_OPENSSL
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#pragma clang diagnostic ignored "-Wused-but-marked-unused"
+#endif
+
+void context::global_init()
+{
+	// this is needed for openssl < 1.0 to decrypt keys created by openssl 1.0+
+#if !defined(OPENSSL_API_COMPAT) || (OPENSSL_API_COMPAT < 0x10100000L)
+	OpenSSL_add_all_algorithms();
+#else
+	OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS | OPENSSL_INIT_ADD_ALL_DIGESTS, nullptr);
+#endif
+}
+
+bool context::verify_dn(std::string const& expected, bool preverified, boost::asio::ssl::verify_context& ctx) {
```

Comment:
This function is specific to running bittorrent over SSL. I don't think it belongs as a member function of the SSL context. I think it would make more sense for it to be a free function (or probably two free functions, on for OpenSSL and one for GnuTLS). But I would think `torrent.cpp` is a pretty reasonable place for it still.

Reply:
I'll move them there then.